tool/teleport-update/main.go

```diff
@@ -0,0 +1,351 @@
+
+	app := libutils.InitCLIParser("teleport-updater", appHelp).Interspersed(false)
+	app.Flag("debug", "Verbose logging to stdout.").Short('d').BoolVar(&ccfg.Debug)
+	app.Flag("data-dir", "Directory to store teleport versions. Access to this directory should be limited.").StringVar(&ccfg.DataDir)
+	app.Flag("log-format", "Controls the format of output logs. Can be `json` or `text`. Defaults to `text`.").Default(libutils.LogFormatText).
+		EnumVar(&ccfg.LogFormat, libutils.LogFormatJSON, libutils.LogFormatText)
+
+	app.HelpFlag.Short('h')
+
+	versionCmd := app.Command("version", "Print the version of your teleport-updater binary.")
+
+	enableCmd := app.Command("enable", "Enable agent auto-updates and perform initial updates.")
+	enableCmd.Flag("proxy", "Address of the Teleport Proxy.").Short('p').Envar(proxyServerEnvVar).StringVar(&ccfg.ProxyServer)
+	enableCmd.Flag("group", "Update group, for staged updates.").Short('g').Envar(updateGroupEnvVar).StringVar(&ccfg.Group)
+	enableCmd.Flag("template", "Go template to override Teleport tgz download URL.").Short('t').Envar(templateEnvVar).StringVar(&ccfg.Template)
+
+	disableCmd := app.Command("disable", "Disable agent auto-updates.")
+
+	updateCmd := app.Command("update", "Update agent to the latest version, if a new version is available.")
+
+	libutils.UpdateAppUsageTemplate(app, args)
+	command, err := app.Parse(args)
+	if err != nil {
+		app.Usage(args)
+		return trace.Wrap(err)
+	}
+	// Logging must be configured as early as possible to ensure all log
+	// message are formatted correctly.
+	if err := setupLogger(ccfg.Debug, ccfg.LogFormat); err != nil {
+		return trace.Errorf("failed to set up logger")
+	}
+
+	if err := ccfg.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	switch command {
+	case enableCmd.FullCommand():
+		err = cmdEnable(ctx, &ccfg)
+	case disableCmd.FullCommand():
+		err = cmdDisable(ctx, &ccfg)
+	case updateCmd.FullCommand():
+		err = cmdUpdate(ctx, &ccfg)
+	case versionCmd.FullCommand():
+		modules.GetModules().PrintVersion()
+	default:
+		// This should only happen when there's a missing switch case above.
+		err = trace.Errorf("command %q not configured", command)
+	}
+
+	return err
+}
+
+func setupLogger(debug bool, format string) error {
+	level := slog.LevelInfo
+	if debug {
+		level = slog.LevelDebug
+	}
+
+	switch format {
+	case libutils.LogFormatJSON:
+	case libutils.LogFormatText, "":
+	default:
+		return trace.Errorf("unsupported log format %q", format)
+	}
+
+	libutils.InitLogger(libutils.LoggingForDaemon, level, libutils.WithLogFormat(format))
+	return nil
+}
+
+const (
+	updatesVersion = "v1"
+	updatesKind    = "updates"
+)
+
+// UpdatesConfig describes the updates.yaml file schema.
+type UpdatesConfig struct {
+	// Version of the configuration file
+	Version string `yaml:"version"`
+	// Kind of configuration file (always "updates")
+	Kind string `yaml:"kind"`
+	// Spec contains configuration.
+	Spec UpdatesSpec `yaml:"spec"`
```

Comment:
Is this intended to be a resource-like thing? We can be a little more freeform with config files.

Reply:
I strongly agree, but see https://github.com/gravitational/teleport/pull/46418#discussion_r1752266134

Reply:
If we go follow the resource pattern. User-facing fields should be under `spec` (basically the updater config like Enabled, Proxy, Group, CDN template) and computed/machine-owned fields should be under `status` (currently only the active version).

Reply:
Will move to `status`